pkg/controller/master/upgrade/upgrade_controller.go

```diff
@@ -792,101 +795,58 @@ func (h *upgradeHandler) storeVMState(upgrade *harvesterv1.Upgrade) error {
-		},
-	}
-	return h.createOrUpdateSecret(secret)
+	upgrade.Annotations[preUpgradeVMsAnnotation] = string(runningVMsJSON)
+	return nil
 }
 
 // restoreVMState attempts to restore all VMs to their pre-upgrade state.
 // This function is designed for single-node environments. There is no guarantee
 // that all VMs will successfully return to their previous state. If any VM fails to start,
 // the error is logged, but the process continues without interruption.
 func (h *upgradeHandler) restoreVMState(upgrade *harvesterv1.Upgrade) {
+	logFields := logrus.Fields{
+		"namespace": upgrade.Namespace,
+		"name":      upgrade.Name,
+	}
 	// ignore the error here as the method should be no-op if there is
 	// something wrong in isSingleNodeRestoreVM, and we want to proceed
 	// to the next step: cleanup jobs, stop repo vm... after upgrade success/failed.
 	isSingleNodeRestoreVM, _ := h.isSingleNodeRestoreVM(upgrade)
 	if !isSingleNodeRestoreVM {
-		logrus.WithFields(logrus.Fields{
-			"namespace": upgrade.Namespace,
-			"name":      upgrade.Name,
-		}).Info("Skip restore VM")
+		logrus.WithFields(logFields).Info("Skip restore VM")
 		return
 	}
 
-	secret, err := h.secretCache.Get(harvesterSystemNamespace, preUpgradeVMsSecretName)
-	if err != nil {
-		logrus.WithFields(logrus.Fields{
-			"namespace": harvesterSystemNamespace,
-			"name":      preUpgradeVMsSecretName,
-		}).WithError(err).Error("Get secret failed")
-		return
-	}
-
-	vms, ok := secret.Data["vms"]
-	if !ok {
-		logrus.WithFields(logrus.Fields{
-			"namespace": harvesterSystemNamespace,
-			"name":      preUpgradeVMsSecretName,
-		}).Error("There is no \"vms\" entry in secret data")
+	if upgrade.Annotations == nil || upgrade.Annotations[preUpgradeVMsAnnotation] == "" {
+		logrus.WithFields(logFields).Infof("Skip restore VM state since %s is not set", preUpgradeVMsAnnotation)
 		return
 	}
 
 	preUpgradeRunningVMs := []PreUpgradeRunningVM{}
-	err = json.Unmarshal(vms, &preUpgradeRunningVMs)
+	err := json.Unmarshal([]byte(upgrade.Annotations[preUpgradeVMsAnnotation]), &preUpgradeRunningVMs)
 	if err != nil {
-		logrus.WithFields(logrus.Fields{
-			"namespace": harvesterSystemNamespace,
-			"name":      preUpgradeVMsSecretName,
-		}).WithError(err).Error("Failed to unmarshal json")
+		logrus.WithFields(logFields).WithError(err).Error("Failed to unmarshal json to PreUpgradeRunningVM list")
 		return
 	}
 
-	for _, preUpgradeRunningVM := range preUpgradeRunningVMs {
-		vm, err := h.vmCache.Get(preUpgradeRunningVM.Namespace, preUpgradeRunningVM.Name)
+	for _, vmInfo := range preUpgradeRunningVMs {
+		vm, err := h.vmCache.Get(vmInfo.Namespace, vmInfo.Name)
 		if err != nil {
-			logrus.WithFields(logrus.Fields{
-				"namespace": preUpgradeRunningVM.Namespace,
-				"name":      preUpgradeRunningVM.Name,
-			}).WithError(err).Error("Failed to get VM from cache")
+			logrus.WithFields(logFields).WithError(err).Errorf("Failed to get VM %s/%s from cache", vmInfo.Namespace, vmInfo.Name)
```

Comment:
If the error happens, we skip this VM. This means we might not ensure the Running VM can Run after the upgrade (somehow get VM failure). Is that right?

If yes, does this assumption meet our expectations?

Reply:
Ideally, there should be a retry mechanism in place. However, the current implementation lacks this, and the restoreVM process is intertwined with image cleanup, making it difficult to handle retries effectively. I'll refactor the entire post-upgrade process as part of [issue #6670](https://github.com/harvester/harvester/issues/6670).

As for your question—yes, this behavior is expected. If something goes wrong, the error is ignored, and the process moves on to start the next VM.

Reply:
Thanks for the explanation!